File: moling-server/app/dao/generation_dao.py
"""澧ㄧ伒 (Moling) 鈥?Generation Task DAO."""

from __future__ import annotations

from typing import Optional

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.dao.base_dao import BaseDAO
from app.models.generation_task import GenerationTask


class GenerationDAO(BaseDAO[GenerationTask]):
    """Data access for GenerationTask records.

    Note: GenerationTask uses a UUID string as primary key, so the
    inherited ``get(id)`` works with string IDs.
    """

    def __init__(self) -> None:
        super().__init__(GenerationTask)

    async def get_by_project(
        self,
        db: AsyncSession,
        project_id: int,
        skip: int = 0,
        limit: int = 50,
    ) -> list[GenerationTask]:
        """List generation tasks for a project, newest first."""
        stmt = (
            select(GenerationTask)
            .where(GenerationTask.project_id == project_id)
            .order_by(GenerationTask.created_at.desc())
            .offset(skip)
            .limit(limit)
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_by_chapter(
        self,
        db: AsyncSession,
        project_id: int,
        chapter_id: int,
    ) -> list[GenerationTask]:
        """List generation tasks for a specific chapter."""
        stmt = (
            select(GenerationTask)
            .where(
                GenerationTask.project_id == project_id,
                GenerationTask.chapter_id == chapter_id,
            )
            .order_by(GenerationTask.created_at.desc())
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_by_status(
        self,
        db: AsyncSession,
        status: str,
        limit: int = 20,
    ) -> list[GenerationTask]:
        """List generation tasks with a specific status (for worker polling)."""
        stmt = (
            select(GenerationTask)
            .where(GenerationTask.status == status)
            .order_by(GenerationTask.created_at.asc())
            .limit(limit)
        )
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def get_by_chapter_and_type(
        self,
        db: AsyncSession,
        chapter_id: int,
        task_type: str,
    ) -> Optional[GenerationTask]:
        """Get the latest generation task for a chapter with a specific type."""
        stmt = (
            select(GenerationTask)
            .where(
                GenerationTask.chapter_id == chapter_id,
                GenerationTask.task_type == task_type,
            )
            .order_by(GenerationTask.created_at.desc())
            .limit(1)
        )
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def get_by_id(
        self,
        db: AsyncSession,
        task_id: str,
    ) -> Optional[GenerationTask]:
        """Get a generation task by ID."""
        return await self.get(db, task_id)

File: moling-server-rs/crates/moling-db/src/dao/generation_dao.rs
//! Generation DAO — generation task CRUD and status-based queries.

use moling_core::error::{AppError, AppResult};
use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, QueryFilter, QueryOrder, QuerySelect};
use uuid::Uuid;

use crate::entities::generation_task::{self, ActiveModel, Entity as GenerationTask, Model as GenerationTaskModel};

/// Generation task data access object.
#[derive(Clone, Default)]
pub struct GenerationDao;

impl GenerationDao {
    /// Find a generation task by ID (UUID string).
    pub async fn find_by_id(
        &self,
        db: &DatabaseConnection,
        id: &str,
    ) -> AppResult<Option<GenerationTaskModel>> {
        let uuid = Uuid::parse_str(id).map_err(|e| {
            tracing::warn!(%id, "Generation: invalid UUID format: {e}");
            AppError::bad_request("Invalid task ID format")
        })?;
        GenerationTask::find_by_id(uuid)
            .filter(generation_task::Column::IsDeleted.eq(false))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "Generation: database error finding by id: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List generation tasks for a project, newest first, with pagination.
    pub async fn list_by_project(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        skip: u64,
        limit: u64,
    ) -> AppResult<Vec<GenerationTaskModel>> {
        GenerationTask::find()
            .filter(generation_task::Column::ProjectId.eq(project_id))
            .filter(generation_task::Column::IsDeleted.eq(false))
            .order_by_desc(generation_task::Column::CreatedAt)
            .offset(skip)
            .limit(limit)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, "Generation: database error listing by project: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List generation tasks for a specific chapter.
    pub async fn list_by_chapter(
        &self,
        db: &DatabaseConnection,
        project_id: i32,
        chapter_id: &str,
    ) -> AppResult<Vec<GenerationTaskModel>> {
        GenerationTask::find()
            .filter(generation_task::Column::ProjectId.eq(project_id))
            .filter(generation_task::Column::ChapterId.eq(chapter_id))
            .filter(generation_task::Column::IsDeleted.eq(false))
            .order_by_desc(generation_task::Column::CreatedAt)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(project_id, %chapter_id, "Generation: database error listing by chapter: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List generation tasks with a specific status (for worker polling).
    pub async fn list_by_status(
        &self,
        db: &DatabaseConnection,
        status: &str,
        limit: u64,
    ) -> AppResult<Vec<GenerationTaskModel>> {
        GenerationTask::find()
            .filter(generation_task::Column::Status.eq(status))
            .filter(generation_task::Column::IsDeleted.eq(false))
            .order_by_asc(generation_task::Column::CreatedAt)
            .limit(limit)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(%status, "Generation: database error listing by status: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Get the latest generation task for a chapter with a specific type.
    pub async fn find_by_chapter_and_type(
        &self,
        db: &DatabaseConnection,
        chapter_id: &str,
        task_type: &str,
    ) -> AppResult<Option<GenerationTaskModel>> {
        GenerationTask::find()
            .filter(generation_task::Column::ChapterId.eq(chapter_id))
            .filter(generation_task::Column::TaskType.eq(task_type))
            .filter(generation_task::Column::IsDeleted.eq(false))
            .order_by_desc(generation_task::Column::CreatedAt)
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%chapter_id, %task_type, "Generation: database error finding by chapter and type: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Create a new generation task.
    pub async fn create(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<GenerationTaskModel> {
        model.insert(db).await.map_err(|e| {
            tracing::error!("Generation: database error creating: {e}");
            AppError::internal("Database insert failed")
        })
    }

    /// Update an existing generation task.
    pub async fn update(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<GenerationTaskModel> {
        model.update(db).await.map_err(|e| {
            tracing::error!("Generation: database error updating: {e}");
            AppError::internal("Database update failed")
        })
    }

    /// Soft-delete a generation task.
    pub async fn soft_delete(&self, db: &DatabaseConnection, id: &str) -> AppResult<()> {
        use sea_orm::{IntoActiveModel, Set};
        use chrono::Utc;

        let entity = self.find_by_id(db, id).await?.ok_or_else(|| {
            AppError::generation_task_not_found()
        })?;

        let mut active = entity.into_active_model();
        active.is_deleted = Set(true);
        active.deleted_at = Set(Some(Utc::now().into()));
        active.update(db).await.map_err(|e| {
            tracing::error!(%id, "Generation: database error soft-deleting: {e}");
            AppError::internal("Database update failed")
        })?;
        Ok(())
    }
}
